src/rgw/rgw_gc.cc

```diff
@@ -65,7 +65,64 @@ int RGWGC::tag_index(const string& tag)
   return rgw_shards_mod(XXH64(tag.c_str(), tag.size(), seed), max_objs);
 }
 
-int RGWGC::send_chain(cls_rgw_obj_chain& chain, const string& tag)
+int RGWGC::send_split_chain(const cls_rgw_obj_chain& chain, const std::string& tag)
+{
+  constexpr int FACTOR_INCREASE = 3; //approximate value by which total size of chain, tag and time increases after encoding into bufferlist
+  ldpp_dout(this, 20) << "RGWGC::send_split_chain - tag is: " << tag << dendl;
+
+  if (cct->_conf->osd_max_write_size) {
```

Comment:
`osd_max_write_size` is an OSD config option, and RGW doesn't have any way to know what that value is on the OSD we're sending this to. this assumes the whole cluster just uses the default value, which might work in most cases, but every OSD in the cluster could _potentially_ have a different value for this

Reply:
so in the event that some OSDs have a lower value, i think we need to handle `OSD_WRITETOOBIG` errors here, by retrying with a lower size like `osd_max_write_size/2`

Reply:
ok yes makes sense, so here cct->_conf->osd_max_write_size, where will the value of osd_max_write_size be picked from? will it be the default value?

Reply:
rgw will see config variables from [client.rgw.xyz], [client], or [global] sections of ceph.conf. if osd_max_write_size isn't specified in any of those, rgw sees its default value

the osds will be looking in sections like [osd.0], [osd] and [global]

Reply:
osd_max_write_size will be present in [global] or the [osd] sections, and not the [client] sections, so rgw will pick this value from the [global] section or the default value will be taken. And if the [osd] sections override these values, and are lesser than them then we will get back an OSD_WRITETOOBIG error, and that is when rgw must retry with osd_max_write_size/2 value. Is my understanding correct?

Reply:
A couple of things here.  max/2 isn't a fixed value in this scenario right?  we might need to retry multiple times, depending on the values chosen at the osd, I think?  I would like to avoid this sort of error handling.  Is there any way to robustify the values in advance?  I thought of, first, just  taking the lower bound value.  Your comment above makes me wonder, is there a reason why we can't know from config system what values the OSD uses?  if not, um, could we allow RGW to ask the osd once up front what it's max is?

Reply:
> Is my understanding correct?

@pritha-srivastava that's right

Reply:
> is there a reason why we can't know from config system what values the OSD uses? if not, um, could we allow RGW to ask the osd once up front what it's max is?

@mattbenjamin this feels like a layering violation to me. the details of crush and its mapping from object names -> PGs -> OSDs are encapsulated by librados/Objecter, and those mappings change as OSDs come up and down

so even if RGW figures out which OSD is the current primary for object 'foo', and fetches that OSD's current `osd_max_write_size` from ceph-mon before sending this ObjectWriteOperation, that OSD may go down in the meantime, and Objecter will retarget the request to another OSD

the OSD's `osd_max_write_size` config could change at any time, so even if RGW does try to cache a value for each OSD, we'd still need a retry strategy that can adapt to config changes

> max/2 isn't a fixed value in this scenario right? we might need to retry multiple times, depending on the values chosen at the osd, I think? I would like to avoid this sort of error handling

agreed, each retry would cut the max length in half until it the OSD returns something other than `OSD_WRITETOOBIG`. ideally, the OSD's `OSD_WRITETOOBIG` reply would also tell us what its limit is. it might be worth exploring an extension to do this, but we'd still need to handle backward compat for OSDs that don't

in the absense of OSD protocol support here, i tend to think that this kind of retry strategy is the best we can do

Reply:
instead of trying to use the OSD's `osd_max_write_size`, maybe RGW's `rgw_max_chunk_size` would be a better default here? we effectively *require* all OSDs to support an `osd_max_write_size` at least that big, or object uploads wouldn't work at all. so we could reasonably expect GC chain requests up to `rgw_max_chunk_size` to succeed, without needing any retry strategy at all

what do you guys think?

Reply:
The default value of rgw_max_chunk_size is 4M and I believe that is considered an optimum value. And if we factor in the bloat caused by encoding, then the max size to send would be ~1M. My concern here is that we would be generating a lot more entries for cls_queue for larger file sizes, although the entries would be smaller in size, but each would have a tag and expiration time - are we going to fill up the queue faster? Infact Matt had suggested that we use the lower bound of osd_max_write_size which is 4M, but it might have the same issue.

Reply:
I strongly prefer a simplified error handling process, and I dont think that limiting single queue entries to 1M (which itself is compensating for an encoding bug) is a serious problem.  I would recommending going with the max-chunk-size, and, if possible, getting someone in RADOS core to look at the encoding problem.